Maze solver and give up feature done.

Ssiyan opened the pull request:
Had to modify some stuff in app.ts to ensure that the feature could be added.
Namely:
Added solveMaze() method inside MazeGame class.
In the MazeGame class, changed 'canvas' in the constructor to be a private variable, so that the new solveMaze() method could access the canvas to create a new MazePath.

Also added a mazeoperations.ts in the 'maze' folder which contains the maze-solving algorithm.

Ssiyan committed:
maze solver done

echau01 commented on src/maze/mazeoperations.ts:
```diff
@@ -0,0 +1,68 @@
+        let curr = toExplore.shift();
+        let neighbors = maze.neighbours(curr)
+
+        for(let c of neighbors) {
```
Style nit: put a space in between the for and the (

echau01 commented on src/maze/mazeoperations.ts:
```diff
@@ -0,0 +1,68 @@
+export function getSolution(maze: Maze, mazePath: MazePath) {
+
+    let visited: boolean[][] = [];     // Keeps track of previously visited paths.
+    let toExplore: Cell[] = [];        // Location data will be stored as an array holding an x and y value.
```
Let's use a queue for toExplore, since array shift() takes O(n) time, while a dequeue should take O(1) time. See: lines 29, 52.

echau01 commented on src/maze/mazeoperations.ts:
```diff
@@ -0,0 +1,68 @@
+                toExplore.push(c);
+            }
+        }
+    }
```
Style nit: leave a blank line between the curly brace and the `let solutionPath` declaration.

echau01 commented on src/maze/mazeoperations.ts:
```diff
@@ -0,0 +1,68 @@
+    let solutionPath: Cell[] = [];
+    toExplore.push(maze.getCell(maze.getRows() - 1, maze.getColumns() - 1));
+
+
```
Style nit: remove the extra blank line.

echau01 commented on src/maze/mazeoperations.ts:
```diff
@@ -0,0 +1,68 @@
+
+        if (!discoveredFrom[currX][currY].equals(maze.getCell(0, 0)))
+            toExplore.push(discoveredFrom[currX][currY]);
+    }
```
Style nit: leave a blank line between the curly brace and the solutionPath.push line.

echau01 commented on src/maze/mazeoperations.ts:
```diff
@@ -0,0 +1,68 @@
+
+// Returns an array of cells which represents the solution to the maze.
+export function getSolution(maze: Maze, mazePath: MazePath) {
+
```
Style nit: remove the extra blank line.

echau01 commented on src/client/app.ts:
```diff
@@ -29,24 +31,39 @@ class MazeGame {
+            const timeDelta: number = finishTime.getTime() - this.startTime.getTime();
+
+            this.isMazeSolved = true;
+            this.mazePath = getSolution(this.maze, new MazePath(new MazeDrawer(this.maze, this.canvas)));
```
I don't know if `getSolution` should return the solution MazePath -- what is the reason for updating `this.mazePath`?

echau01 commented on src/client/app.ts:
```diff
@@ -29,24 +31,39 @@ class MazeGame {
-        const canvas: HTMLCanvasElement = document.getElementById("maze") as HTMLCanvasElement;
-        canvas.width = canvasWidth;
-        canvas.height = canvasHeight;
+        //const canvas: HTMLCanvasElement = document.getElementById("maze") as HTMLCanvasElement;
```
Let's remove the comment.

echau01 commented on src/maze/mazeoperations.ts:
```diff
@@ -0,0 +1,68 @@
+            }
+        }
+    }
+    let solutionPath: Cell[] = [];
```
I have an idea to make the backtracking portion a bit more clean. Notice that a cell can be discovered from at most one cell. Thus, we don't need to use the array toExplore. Let's do something like:

```
let curr = maze.getCell(maze.getRows() - 1, maze.getColumns() - 1);

while (curr) {
    solutionPath.push(curr);
    curr = discoveredFrom[curr.getColumn()][curr.getRow()];
}
```

echau01 commented on src/maze/mazeoperations.ts:
```diff
@@ -0,0 +1,68 @@
+    }
+    solutionPath.push(maze.getCell(0, 0));
+
+    for (let c of solutionPath.reverse()) {
```
Let's replace the for loop with:

```
for (let i = solutionPath.length - 1; i >= 0; i--) {
    mazePath.add(solutionPath[i]);
}
```

This saves an O(n) time reverse() operation.

echau01 reviewed (changes requested):
Thanks for the PR -- I've left some changes to consider.

Ssiyan committed:
did some requested changes

echau01 commented on src/maze/mazeoperations.ts:
```diff
@@ -3,33 +3,31 @@ import Maze from "./maze";
 
-// Returns an array of cells which represents the solution to the maze.
+// Returns a MazePath which represents the solution to the maze.
 export function getSolution(maze: Maze, mazePath: MazePath) {
```
I think we can refactor this function a little bit. This function doesn't return a mazepath; it just draws the solution onto the maze canvas, so I suggest renaming the function to `drawSolution`.

We aren't really using the mazePath that is passed in -- all that's truly being used is the canvas object in the mazepath. We're currently forcing the caller to construct a new mazepath, which is a bit unintuitive. Let's instead create a mazepath _inside_ the function, and replace the `mazePath` parameter with a `canvas` parameter, like so:

```
export function drawSolution(maze: Maze, canvas: HTMLCanvasElement)
```

Then in `app.ts`, we can simply pass in the `this.canvas` object of the MazeGame class. We make a new MazePath to store the solution _inside_ the `drawSolution` function. This way, we can hide implementation details from client code.

```
export function drawSolution(maze: Maze, canvas: HTMLCanvasElement) {
    const mazePath = new MazePath(new MazeDrawer(maze, canvas));`
    /* Rest of the code... */
}
```

Finally, update the comment above the `drawSolution` function to say something like:

```
/**
 * Draws the solution of the given maze onto the given canvas.
 */
```

echau01 commented on src/client/app.ts:
```diff
@@ -56,7 +55,8 @@ class MazeGame {
 
             this.isMazeSolved = true;
-            this.mazePath = getSolution(this.maze, new MazePath(new MazeDrawer(this.maze, this.canvas)));
+            //let mazeSolution: MazePath =
```
Let's remove this comment.

echau01 reviewed (changes requested):
Looks good so far. I've suggested some changes.

Ssiyan committed:
Did some further refactoring of mazeoperations.ts and app.ts as requested.

-deleted comment in app.ts
-renamed getSolution to drawSolution and adjusted the comments accordingly
-modified the second parameter value to accept a canvas object and not a mazePath object, and adjusted things accordingly

Ssiyan committed:
Custom queue class has been implemented.

Given that arrays in TypeScript are dynamic, it's a little bit different from a standard queue implementation in that resizing is not handled.

Otherwise it should be O(1) for both pop and push.

echau01 commented:
Suggestion: the memory efficiency can be improved. Every pop operation replaces the front of the queue with `undefined` -- that's what the line `delete this._arr[this.exit];` does. As a result, after n `push` and n `pop` operations, the queue is empty, but the underlying array looks like this:

`[undefined, undefined, undefined, ..., undefined] // n "undefined"s`

After researching online for efficient queue implementations in JavaScript, I think there are two possible solutions:
1. Use a simple linked list implementation.
2. Use "delayed shift arrays", as described [here](https://stackoverflow.com/a/49902604/3335320).

You can decide which one you want to implement.

Ssiyan committed:
Improved memory efficiency by making the array-based queue shift left after >20 pop operations.

echau01 committed:
Implement queue pop with delayed shift queue

echau01 committed:
Beautify comments and types

echau01 reviewed (commented):
Thanks for the great PR!
